### PR TITLE
Ex 84 update forgot password

### PR DESCRIPTION
**Pull request by JMStewart00**
We needed a way to update passwords and also wanted to add some success messaging or denials on login attempts. Hopefully this all looks good. 

Went for using the PasswordResetTokenGenerator from Django to send it to the front end. Parsing it on the front end and sending it back to the backend for validation. Couldn't find a way to make the token expire any less than a single day.

### CODE REVIEW

File: lfucg-exactions/client/js/components/LoginPage.js [hunk 1]
```diff
@@ -102,8 +94,13 @@ function mapDispatchToProps(dispatch) {
         onLogin(event) {
             event.preventDefault();
             dispatch(login())
-            .then(() => {
```
Comment (KellyInnovation) [inline lfucg-exactions/client/js/components/LoginPage.js]: This was being used to return the user to the page they were on when they clicked Login.

File: lfucg-exactions/client/js/components/ResetPassword.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                            <form onSubmit={resetPassword}>
+                                <fieldset>
+                                    <div className="row">
+                                        <div className="col-md-6 col-md-offset-3 col-sm-8 col-sm-offset-2">
+                                            <FormGroup label="Enter Password" id="password_1">
+                                                <input
+                                                  type="password"
+                                                  className={`form-control ${activeForm.password_1 !== activeForm.password_2 ? 'error' : ''}`}
+                                                  placeholder="Enter Password"
+                                                  aria-label="Password Enter"
```
Comment (KellyInnovation) [inline lfucg-exactions/client/js/components/ResetPassword.js]: aria labels and other labels should match (there should be one best label)

File: lfucg-exactions/client/js/components/ResetPassword.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                                        </div>
+                                    </div>
+                                    <div className="row">
+                                        <div className="col-md-6 col-md-offset-3 col-sm-8 col-sm-offset-2">
+                                            <FormGroup label="Confirm Password" id="password_2">
+                                                <input
+                                                  type="password"
+                                                  className={`form-control ${activeForm.password_1 !== activeForm.password_2 ? 'error' : ''}`}
+                                                  placeholder="Confirm Password"
+                                                  aria-label="Password Confirm"
```
Comment (KellyInnovation) [inline lfucg-exactions/client/js/components/ResetPassword.js]: See above for labels

File: lfucg-exactions/client/js/components/LoginPage.js (hunk 1 above)
Comment (JMStewart00) [reply lfucg-exactions/client/js/components/LoginPage.js]: It's going to redirect them directly back to the reset password page unless we find a better option.

File: lfucg-exactions/server/accounts/auth.py [hunk 4]
```diff
@@ -97,11 +99,11 @@ def post(self, request, *args, **kwargs):
 @api_view(['POST'])
 @permission_classes((AllowAny, ))
 def forgot_password(request):
-    email = request.data.get('email', None)
-    user = User.objects.filter(email__icontains=email)
-    if user.exists():
-        send_password_reset_email(user)
-        
+    email = request.data.get('email_forgot_password', None)
+    user = User.objects.get(email=email)
```
Comment (KellyInnovation) [inline lfucg-exactions/server/accounts/auth.py]: Check first to make sure email is not None

Comment (KellyInnovation) [reply lfucg-exactions/server/accounts/auth.py]: What happens if there is more than one user with that email?

File: lfucg-exactions/server/accounts/auth.py [hunk 5]
```diff
@@ -117,30 +119,28 @@ def forgot_username(request):
...
+        _delete_token(token)
+        response = {'success': 'New password has been saved.'}
+
+    return Response(response, status=status)
+
+def _validate_token(user, token):
+    response = {}
+    status = statuses.HTTP_400_BAD_REQUEST
+    if PasswordResetTokenGenerator().check_token(user, token):
+        print('we outchea')
```
Comment (KellyInnovation) [inline lfucg-exactions/server/accounts/auth.py]: Remove printe statement

File: lfucg-exactions/server/accounts/auth.py (hunk 4 above)
Comment (JMStewart00) [reply lfucg-exactions/server/accounts/auth.py]: Do we still have to check if it is not none if that's what is being sent from the front end with messaging that says "if there is an email that matches we'll send you instructions"?

Also didn't test more than one person with that email. I can put in a safeguard for that... Are emails unique?

Comment (KellyInnovation) [reply lfucg-exactions/server/accounts/auth.py]: Yes, we still need to test for if it is None.  Many checks like that are to prevent the system or users misbehaving.  
Emails are likely unique, but we are not putting any unique email requirements in the system, so there could be a repeat.

File: lfucg-exactions/client/js/components/ResetPassword.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                                        <div className="col-md-6 col-md-offset-3 col-sm-8 col-sm-offset-2">
+                                            <FormGroup label="Confirm Password" id="password_2">
+                                                <input
+                                                  type="password"
+                                                  className={`form-control ${activeForm.password_1 !== activeForm.password_2 ? 'error' : ''}`}
+                                                  placeholder="Confirm Password"
+                                                  aria-label="Confirm Password"
+                                                />
+                                            </FormGroup>
+                                            <p className={`help-block bg-danger text-center ${activeForm.password_1 !== activeForm.password_2 ? '' : 'hidden'}`} aria-label="Password Mismatch Warning">
```
Comment (Wei-KeatKong) [inline lfucg-exactions/client/js/components/ResetPassword.js]: Try to have your aria labels match your descriptive text

File: lfucg-exactions/server/server/settings/base.py [hunk 7]
```diff
@@ -310,4 +310,5 @@ def custom_show_toolbar(request):
 #CEF_VENDOR = 'Your Company'
 #CEF_VERSION = '0'
 #CEF_DEVICE_VERSION = '0'
-PAGINATION_SIZE = 10
\ No newline at end of file
+PAGINATION_SIZE = 10
+PASSWORD_RESET_TIMEOUT_DAYS = 1
```
Comment (Wei-KeatKong) [inline lfucg-exactions/server/server/settings/base.py]: Check if they are okay with the reset links being active for a day and the token being deleted after a successful password change. If not, you can read this for ideas. http://pminkov.github.io/blog/how-to-generate-user-activation-links-in-django.html

Comment (JMStewart00) [reply lfucg-exactions/server/server/settings/base.py]: Ok thanks! We are checking with Craig and will see if that is good or not.

**Comment by JMStewart00** [comment]
Merging with the caveat that we'll go back in if they want password resets to be valid for shorter amount of time. User testing wanted some of the features available so we're moving forward per conversation with @KellyInnovation!
